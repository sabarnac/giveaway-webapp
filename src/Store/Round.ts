import { observable, computed } from "mobx";
import Participant from "./Participant";
import Config from "./Config/Config";
import chunk from "lodash.chunk";
import Match from "./Match";

/**
 * Class containing the details of a round of the tournament.
 */
export default class Round {
  @observable public matches: Match[];

  private _config: Config;

  private _shuffleParticipants = (participants: Participant[]): Participant[] =>
    this._config.randomGenerator.shuffle(participants);

  private _createMatch = (participants: Participant[]): Match =>
    new Match(this._config, participants);

  private _getMatches = (participants: Participant[]): Match[] =>
    chunk(
      this._shuffleParticipants(participants),
      this._config.participantsPerMatch,
    ).map(this._createMatch);

  public constructor(config: Config, participants: Participant[]) {
    this._config = config;

    this.matches = this._getMatches(participants);
  }

  private _getMatchParticipants = (match: Match): Participant[] =>
    match.participants;

  private _getMatchWinner = (match: Match): Participant => match.winner;

  private _getMatchLosers = (match: Match): Participant[] => match.losers;

  /**
   * Get the list of participants in the round.
   * @returns {Participant[]} The list of participants.
   */
  @computed public get participants(): Participant[] {
    return this.matches.map(this._getMatchParticipants).flat();
  }

  /**
   * Get the list of winners in the round.
   * @returns {Participant[]} The list of winners.
   */
  @computed public get winners(): Participant[] {
    return this.matches.map(this._getMatchWinner);
  }

  /**
   * Get the list of losers in the round.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this.matches.map(this._getMatchLosers).flat();
  }
}
