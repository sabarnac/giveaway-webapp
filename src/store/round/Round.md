Usage:

```tsx
import Match from "./store/round/match/Match";
import Participant from "./store/round/match/participant/Participant";

const round: Round = new Round(Config.instance, [
  new Participant("FOO"),
  new Participant("BAR")
]);
const roundId: Round = round.id; // "round-1"
const matches: Match[] = round.matches; // [Match("match-1"), ..., Match("match-x")]
const firstMatch: Match = round.firstMatch; // Match("match-1")
const lastMatch: Match = round.lastMatch; // Match("match-x")
const participants: Participant[] = round.participants; // [Participant("FOO"), Participant("BAR")]
const winners: Participant[] = round.winners; // [Participant("FOO")]
const losers: Participant[] = round.losers; // [Participant("BAR")]
```
