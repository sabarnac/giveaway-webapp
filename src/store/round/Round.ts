import { observable, computed } from "mobx";
import Participant from "./match/participant/Participant";
import Config from "../config/Config";
import chunk from "lodash.chunk";
import Match from "./match/Match";
import last from "lodash.last";
import RandomGenerator from "../config/RandomGenerator";

/**
 * Class containing the details of a round of the tournament.
 */
export default class Round {
  @observable private _id: string;
  @observable private _matches: Match[];

  private static counter: number = 1;

  private _config: Config;

  private _shuffleParticipants = (participants: Participant[]): Participant[] =>
    RandomGenerator.shuffle(participants);

  private _createMatch = (participants: Participant[]): Match =>
    new Match(participants);

  private _getMatches = (participants: Participant[]): Match[] =>
    chunk(
      this._shuffleParticipants(participants),
      this._config.participantsPerMatch
    ).map(this._createMatch);

  public constructor(config: Config, participants: Participant[]) {
    this._config = config;

    this._id = `round-${Round.counter++}`;
    this._matches = this._getMatches(participants);
  }

  private _getMatchParticipants = (match: Match): Participant[] =>
    match.participants;

  private _getMatchWinner = (match: Match): Participant => match.winner;

  private _getMatchLosers = (match: Match): Participant[] => match.losers;

  /**
   * Get the match ID.
   * @returns {string} The unique ID of the match.
   */
  @computed public get id(): string {
    return this._id;
  }

  /**
   * Get the list of matches in the round.
   * @returns {Match[]} The list of matches.
   */
  @computed public get matches(): Match[] {
    return this._matches;
  }

  /**
   * Get the first match of the round.
   * @returns {Match} The first match.
   */
  @computed public get firstMatch(): Match {
    return this._matches[0];
  }

  /**
   * Get the last match of the round.
   * @returns {Match} The last match.
   */
  @computed public get lastMatch(): Match {
    return <Match>last(this._matches);
  }

  /**
   * Get the list of participants in the round.
   * @returns {Participant[]} The list of participants.
   */
  @computed public get participants(): Participant[] {
    return this._matches.map(this._getMatchParticipants).flat();
  }

  /**
   * Get the list of winners in the round.
   * @returns {Participant[]} The list of winners.
   */
  @computed public get winners(): Participant[] {
    return this._matches.map(this._getMatchWinner);
  }

  /**
   * Get the list of losers in the round.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this._matches.map(this._getMatchLosers).flat();
  }
}
