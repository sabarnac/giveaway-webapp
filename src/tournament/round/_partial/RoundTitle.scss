.round > h2 {
  color: #ffffff;
}
