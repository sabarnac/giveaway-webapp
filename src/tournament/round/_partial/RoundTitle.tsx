import { withTranslation } from "react-i18next";
import RoundTitle from "./RoundTitle.react";
/**
 * React component for the round title.
 */
export default withTranslation()(RoundTitle);
