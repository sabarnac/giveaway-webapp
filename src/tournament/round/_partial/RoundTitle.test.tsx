import React from "react";
import RoundTitle from "./RoundTitle.react";
import Round from "../../../store/round/Round";
import { create, ReactTestRenderer } from "react-test-renderer";
import {
  createDummyRound,
  createDummyTranslationProps,
} from "../../../util/test";

describe("RoundTitle Component.", () => {
  it("Roundes snapshot.", () => {
    const round: Round = createDummyRound();
    const component: ReactTestRenderer = create(
      <RoundTitle round={round} {...createDummyTranslationProps()} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
