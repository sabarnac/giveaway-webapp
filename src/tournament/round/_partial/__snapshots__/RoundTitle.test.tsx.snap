// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoundTitle Component. Roundes snapshot. 1`] = `
<h2>
  Round 
  1
</h2>
`;
